peps/pep-0742.rst

```diff
@@ -0,0 +1,305 @@
+exactly that.
+
+Other examples of issues that arose out of the current behavior of ``TypeGuard`` include:
+
+* `Python typing issue <https://github.com/python/typing/issues/996>`__ (``numpy.isscalar``)
+* `Python typing issue <https://github.com/python/typing/issues/1351>`__ (:py:func:`dataclasses.is_dataclass`)
+* `Pyright issue <https://github.com/microsoft/pyright/issues/3450>`__ (expecting :py:data:`typing.TypeGuard` to work like :py:func:`isinstance`)
+* `Pyright issue <https://github.com/microsoft/pyright/issues/3466>`__ (expecting narrowing in the ``else`` branch)
+* `Mypy issue <https://github.com/python/mypy/issues/13957>`__ (expecting narrowing in the ``else`` branch)
+* `Mypy issue <https://github.com/python/mypy/issues/14434>`__ (combining multiple TypeGuards)
+* `Mypy issue <https://github.com/python/mypy/issues/15305>`__ (expecting narrowing in the ``else`` branch)
+* `Mypy issue <https://github.com/python/mypy/issues/11907>`__ (user-defined function similar to :py:func:`inspect.isawaitable`)
+* `Typeshed issue <https://github.com/python/typeshed/issues/8009>`__ (``asyncio.iscoroutinefunction``)
+
+Rationale
+=========
+
+The problems with the current behavior of :py:data:`typing.TypeGuard` compel us to improve
+the type system to allow a different type narrowing behavior. :pep:`724` proposed to change
+the behavior of the existing :py:data:`typing.TypeGuard` construct, but we :ref:`believe <pep-742-change-typeguard>`
+that the backwards compatibility implications of that change are too severe. Instead, we propose
+adding a new special form with the desired semantics.
+
+We acknowledge that this leads to an unfortunate situation where there are two constructs with
+a similar purpose and similar semantics. We believe that users are more likely to want the behavior
+of ``TypeNarrower``, the new form proposed in this PEP, and therefore we recommend that documentation
+emphasize ``TypeNarrower`` over ``TypeGuard`` as a more commonly applicable tool. However, the semantics of
+``TypeGuard`` are occasionally useful, and we do not propose to deprecate or remove it. In the long
+run, most users should use ``TypeNarrower``, and ``TypeGuard`` should be reserved for rare cases
+where its behavior is specifically desired.
+
+
+Specification
+=============
+
+A new special form, ``TypeNarrower``, is added to the :py:mod:`typing`
+module. Its usage, behavior, and runtime implementation are similar to
+those of :py:data:`typing.TypeGuard`.
+
+It accepts a single
+argument and can be used as the return type of a function. A function annotated as returning a
+``TypeNarrower`` is called a type narrowing function. Type narrowing functions must return ``bool``
+values, and the type checker should verify that all return paths return
+``bool``. However, ``TypeNarrower`` is not a subtype of ``bool``.
+The type ``Callable[..., TypeNarrower[int]]`` is not assignable to
+``Callable[..., bool]`` or ``Callable[..., TypeGuard[int]]``, and vice versa.
```

Comment:
I think it would be valuable for the PEP to include a rationale for this. I think it was really unfortunate that PEP 647 included this restriction without providing any rationale. There may be good reasons for this restriction (I would guess maybe it has to do with losing the benefits -- and maybe even back-compat implications? -- of having bool no longer be a final type?), but they are not obvious and so they should be clearly recorded. It seems useful to be able to pass a type-narrower function as a predicate callback (whose signature would be `Callable[..., bool]`, so the burden of proof should be on the rationale for not allowing it.

Reply:
I'm not sure what the rationale was here; maybe @erictraut remembers why this restriction was put in. In any case, if we change this, we'd likely want to change it for both TypeGuard and this PEP's new construct, and we could probably do that with a smaller (non-PEP) modification of the spec.

Reply:
IIRC, this statement wasn't in my early drafts of PEP 647 and was added near the end at the suggestion of @gvanrossum, who sponsored the PEP. I agree that the restriction should either be removed or a better rationale or formulation should be provided.

Reply:
python/mypy#11307 might be part of the reason, though that issue postdates the acceptance of the PEP.

Reply:
FYI I opened https://discuss.python.org/t/compatibility-between-typeguard-and-bool/46364 about this point.